### PR TITLE
[WIP] Jamiecook/strtree3

### PR DESCRIPTION
**Pull request by jamiecook**
Had to create a new branch in order to merge upstream master... too many merge conflicts

### CODE REVIEW

**Comment by ewestern** [comment]
I have some additions to this branch, but it looks like some tests are failing due to issues that *might* be resolved by my other branches. 
I'll merge those in so hopefully I can update this branch.

**Comment by jamiecook** [comment]
I've just  rebased this on top of latest master from ewestern/geos and am still getting some errors when testing... not sure if it's related to your other work or just that my implementation is flawed

**Comment by ewestern** [comment]
There are definitely still some errors, but the str functions weren't working beforehand on this branch (At least not when using more than a few geometries).
I've gotten rid of some of the errors, but it looks like there is still a case of a geometry (or other object) escaping the scope of a with block. Still need to track that down.

**Comment by ewestern** [comment]
Ok, I think I figured out the issue with STRTree and ... it's not looking great.
As near as I can tell, we'll need a completely different API. The issue is knowing when to deallocate the pointers to the arbitrary values we're mapping to (In our test cases: the Ints).
There's no real way to know when to deallocate them if we expose a pure API.

I sketched out an option here:
https://gist.github.com/ewestern/d9e138a4f2a489e99925abbe1c8f39ad

Really not thrilled about that, but not sure how else to handle it at the moment. Will have to think on it.

**Comment by jamiecook** [comment]
I think i see, each time we insert, we're allocating a bit more memory - the gist you sketched basically wraps up that list and carries it around with the STRTree so that when that object goes out of scope they can be cleaned up... about right?

**Comment by jamiecook** [comment]
Are there not other ways of allocating the memory for the storables that don't immediately destroy it?

Reading the description of alloca it seems to be intended for short term pointer needs
image

Maybe good ole malloc to the rescue?
image

**Comment by ewestern** [comment]
So, I was actually already using alloca, (here: https://github.com/jamiecook/geos/blob/jamiecook/strtree3/src/Data/Geometry/Geos/Raw/STRTree.hs#L34)  but that was the problem: The pointer would be marked for deallocation as soon as the insert was complete. (As such, when we queried the tree afterwards, it would sometimes return random values that happened to be in the same position on the heap). But what we want is for that pointer to last for the lifetime of the tree object (and no longer!).
For malloc, the issue is knowing when the tree is going to be collected and thus the pointer itself can be deallocated.

However ... I do think I found a way to do it without changing the API. It required using the Foreign.Concurrent library. Will push some commits.

File: src/Data/Geometry/Geos/STRTree.hs [hunk 1]
```diff
@@ -1,7 +1,52 @@
...
+toVector :: Storable a => RT.STRTree a -> V.Vector a
+toVector = foldr V.cons V.empty
+
+-- would like to expose 'empty' as is common for haskell collections, but when initializing an STRTree we have to know the node size before hand
+
+fromList :: Storable b => [(Geometry a, b)] -> RT.STRTree b
+fromList = fromFoldable
+
+
+insert :: Storable a => Geometry b -> a -> RT.STRTree a -> RT.STRTree a
```
Comment (ewestern) [inline src/Data/Geometry/Geos/STRTree.hs]: Actually, this shouldn't be exposed in the main API. 
From the docs:

> However, once the tree has been built (explicitly or on the first call to query), items may not be added or removed.

File: src/Data/Geometry/Geos/STRTree.hs [hunk 2]
```diff
@@ -1,7 +1,52 @@
...
+insert geom item tree = runGeos $ do
+    rg :: RG.GeomConst <- convertGeometryToRaw geom
+    RT.insert tree rg item
+    return tree
 
+fromFoldable :: (Foldable f, Storable b) => f (Geometry a, b) -> RT.STRTree b
+fromFoldable things = runGeos $ do
+  tree <- RT.createSTRTree $ max 4 $ length things
```
Comment (ewestern) [inline src/Data/Geometry/Geos/STRTree.hs]: The designation for nodeCapacity here is tricky.
It really shouldn't be "a node for each of the items," but I was definitely running out of capacity when running the tests with a default size of 10.
Will have to think on how to handle this.

File: src/Data/Geometry/Geos/STRTree.hs (hunk 1 above)
Comment (jamiecook) [reply src/Data/Geometry/Geos/STRTree.hs]: good call, just a generic from foldable/traversable method which builds it in one hit

File: src/Data/Geometry/Geos/STRTree.hs (hunk 2 above)
Comment (jamiecook) [reply src/Data/Geometry/Geos/STRTree.hs]: I don't think this is a capacity issue, my reading of the guidance for nodeCapacity is that it's the per node capacity before a node will be split (or whatever the STR equivalent of node splitting is). So changing this number will change the shape of your tree (low will make it tall but skinny, a big number will make it wide but short). 

And making it equal to "number of items to store" will result in a tree with a single node containing each of the items added...

So the errors  you are seeing are probably related to something else I'm afraid :/

File: tests/SpatialOperationsSpec.hs [hunk 3]
```diff
@@ -66,3 +71,41 @@ spatialOpsSpecs = describe "Tests Contains" $ do
...
+    polygons <- (fmap ensurePolygon) <$> loadThingsFromFile "tests/sampledata/polygons.csv"
+    -- count <- timeIt $! layerIntersect polygons points -- 23.4 seconds
+    let count = 176
+    count `shouldBe` 176
+    count' <- timeIt $ layerIntersectSTR polygons points
+    count' `shouldBe` count
+    count'' <- timeIt $ layerIntersectSTR' points polygons
+    count'' `shouldBe` count'
+
+layerIntersect :: [Geometry Polygon] -> [Geometry Point] -> IO Int
```
Comment (jamiecook) [inline tests/SpatialOperationsSpec.hs]: So this is interesting, I've written the dumb layer intersection code here - for each style. 

The resulting count is 110, not 176 as is coming back from the other two versions using the STR trees

The timing is good though - 24seconds compared to 0.1/0.5 seconds

File: src/Data/Geometry/Geos/Raw/Geometry.hs [hunk 4]
```diff
@@ -191,32 +191,34 @@ getNumInteriorRings :: Geometry a => a -> Geos Int
...
       => (I.GEOSContextHandle_t -> Ptr I.GEOSGeometry -> CInt -> IO (Ptr I.GEOSGeometry))
       -> a
       -> Int
-      -> Geos b
+      -> Geos GeomConst
 getN_ f g i = 
   withGeos $ \h ->  withGeometry g $ \gp ->  do
       gp' <- throwIfNull "getN" $ f h gp $ fromIntegral i
-      constructGeometry h gp'
+      I.geos_GeomClone h gp' >>= constructGeometry h
```
Comment (jamiecook) [inline src/Data/Geometry/Geos/Raw/Geometry.hs:206]: what is the logic here with using clone in all these methods?

File: tests/SpatialOperationsSpec.hs [hunk 5]
```diff
@@ -66,3 +71,35 @@ spatialOpsSpecs = describe "Tests Contains" $ do
...
+    count' <- timeIt $ layerIntersectSTR polygons points
+    count' `shouldBe` count
+
+layerIntersect :: [Geometry Polygon] -> [Geometry Point] -> IO Int
+layerIntersect polygons points = do
+  let things = sum $ fmap (f points) polygons
+  if (things > 0) then pure things else pure 0
+  where f points polygon = length $ filter (\p -> contains polygon p) points
+
+layerIntersectSTR :: [Geometry Polygon] -> [Geometry Point] -> IO Int
```
Comment (jamiecook) [inline tests/SpatialOperationsSpec.hs:99]: I've included the actual implementation of layer intersection using the STR tree here as it's not obvious that a query isn't the same as an intersect/contains query... but the numbers line up with the dumb impl so... thats something anyways :)

**Comment by ewestern** [comment]
Added some cleanup on https://github.com/ewestern/geos/pull/23
Closing for that one.

**Pull request closed without merging**